src/Lib.hs

```diff
@@ -23,31 +23,36 @@ import Base.Interval as I
     , intervalToDistance
     )
 import Scale (Scale(..), BaseMode(..), baseModeIntervals)
-import Data.Set(Set(..))
+
+import Control.Monad (foldM)
+import Data.Set (Set(..))
 import qualified Data.Set as S
 import Data.Map.Strict (Map, insert, fromList, toList, (!), delete, (!?))
 import Data.Maybe (fromJust)
 
 
-chordToNotes :: Chord -> [Root]
+chordToNotes :: Chord -> Maybe [Root]
 chordToNotes chord =
-  flip jumpIntervalFromNote (getChordRoot chord) <$> S.toList (chordToIntervals chord)
+  do chordInts <- chordToIntervals chord
+     return $ flip jumpIntervalFromNote (getChordRoot chord) <$> S.toList chordInts
 
 
-chordToIntervals :: Chord -> Set Interval
+chordToIntervals :: Chord -> Maybe (Set Interval)
 chordToIntervals chord =
-  let
-    baseScale = highestNaturalToIntervals (getHighestNatural chord) $ qualityToIntervals $ C.getQuality chord
-    intervals = susIntervals (extendIntervals baseScale $ getExtensions chord) $ getSus chord
-  in
-    foldr S.insert S.empty intervals
+  do qualInts <- qualityToIntervals $ C.getQuality chord
+     baseScale <- highestNaturalToIntervals (getHighestNatural chord) qualInts
+     extendedScale <- extendIntervals baseScale $ getExtensions chord
+     intervals <- susIntervals extendedScale $ getSus chord
+     return $ foldr S.insert S.empty intervals
 
 
 type HeliotonicScale = Map Int Interval
 
 
-qualityToIntervals :: CQ.Quality -> HeliotonicScale
-qualityToIntervals qual = fromList $ zip [1..7] $ S.toList $ baseModeIntervals $ qualityToScale qual
+qualityToIntervals :: CQ.Quality -> Maybe HeliotonicScale
```

Comment:
Because the chord qualities are limited to just 5 and we come up with the mode/intervals to use for each, don't we have a guarantee that these HeliotonicScales have valid intervals? Do we get anything from making this partial?

Reply:
Same question for the rest of the functions in Lib -- I think the modifications that these three functions are making are tightly constrained enough that they'll always be valid intervals?

For susIntervals, we use baseQuality i with i, so that should always be valid. (As a sidenote, I'm not sure this was changed in #17 to account for the new shape of the Sus datatype.) And for extendIntervals, we do the same baseQuality thing, and the <+> function should also guarantee us a valid interval if one is given to it. For highestNat, we only ever try to insert a M7, which is definitely valid.

Let me know if I'm missing something

Reply:
Just nitpicking terminology: Changing these to `Maybe`s was in an effort to make them _total_ functions.

But you're not wrong about some of this stuff. There are constraints implied by the arguments to these functions, and correctness is only dependent on the implementations provided. I'll do another pass over `Lib` and `Scale`; I should have been more critical about the contracts of these functions since, as you said, there's lots of stuff we know to be true.

I mean, technically there's nothing 'wrong' with making all these functions return `Maybe`s, but it can definitely get annoying if we actually know for a fact that there's no way it can fail.